[GR-42598] Logging to a Native Executable Demo. (#156)
graalvm/graalvm-demos: native-image-logging-examples/README.md

# Configuring Logging in Native Image

> This demo demonstrates how Native Image supports logging, which is using the `java.util.logging.*` API by default. The logger can be initialized at build time with a custom _logging.properties_ configuration file or at runtime. The below examples show both cases.

Reviewer: ```suggestion
This demo demonstrates how the `java.util.logging.*` API can be used with Native Image. The logger can be initialized at build time with a custom _logging.properties_ configuration file or at run time. The below examples show both cases.
```

> This demo demonstrates how the `java.util.logging.*` API can be used with Native Image. If your application require additional logging handlers, you can register them with a custom configuration file, _logging.properties_ , and initialize at build time. This reduces the size of the resulting executable file and improves the startup time. Unless your application needs to process _logging.properties_ at run time, this approach is recommended. Both approaches are demonstrated in the examples. 

Reviewer: "demo demonstrates" is a bit redundant (note I also improved the text a bit):
```suggestion
This demo shows how the `java.util.logging.*` API can be used with Native Image. If your application requires additional logging handlers, you can register them with a custom configuration file, _logging.properties_ , and initialize at build time. This reduces the size of the resulting executable file and improves startup time. Unless your application needs to process _logging.properties_ at run time, this approach is recommended. Both approaches are demonstrated in the examples. 
```

### Prerequisites
* [GraalVM](http://graalvm.org)
* [Native Image](https://www.graalvm.org/docs/reference-manual/native-image/)

## Preparation

1. Download and install the latest GraalVM JDK with Native Image using [GraalVM JDK Downloader](https://github.com/graalvm/graalvm-jdk-downloader):
  ```bash
  bash <(curl -sL https://get.graalvm.org/jdk)
  ```

2. Download or clone the repository and navigate into the `native-image-logging-examples` directory:
  ```bash
  git clone https://github.com/graalvm/graalvm-demos
  cd graalvm-demos/native-image-logging-examples
  ```

There are two Java classes: one for the build-time logger initialization and the second for runtime logger initialization. The logger will be initialized with a custom _logging.properties_ configuration file, which is placed in the same directory as _BuildTimeLoggerInit.java_ and _RuntimeLoggerInit.java_.

> ## Build-Time Logger Initialization

Author: Do you mean rename both  .java file in this demo?

In this example, the logger will be initialized at build time with a custom _logging.properties_ configuration file, placed in the same repository as _BuildTimeLoggerInit.java_.

1. Compile _BuildTimeLoggerInit.java_ using `javac`:

    ```bash

>     javac BuildTimeLoggerInit.java

Reviewer: This also implies that the user has added the GraalVM installation to their path. Maybe be explicit in this and following commands?
```suggestion
    $JAVA_HOME/bin/javac BuildTimeLoggerInit.java
```

Author: Okay , I will change  it in all following commands!

# Native Image Logging Configuration Examples

> 

Reviewer 2: Change the title to: 

**Configuring Logging in Native Image**

This repository contains the code for a demo application for [GraalVM](http://graalvm.org).

### Prerequisites
* [GraalVM](http://graalvm.org)
* [Native Image](https://www.graalvm.org/docs/reference-manual/native-image/)

## Preparation

> 1. [Download GraalVM](https://www.graalvm.org/downloads/), unzip the archive, export the GraalVM home directory as the `$GRAALVM_HOME` and add `$GRAALVM_HOME/bin` to the `PATH` environment variable.

Reviewer 2: Replace step 1 with just this line of text:

1. Download and install the latest GraalVM JDK with Native Image using [GraalVM JDK Downloader](https://github.com/graalvm/graalvm-jdk-downloader):
  ```
  bash <(curl -sL https://get.graalvm.org/jdk)
  ```

>   On Linux:

Reviewer 3: I don't think you need these lines because we can assume the user knows how to do this. Also, it makes assumptions about the location of the directory.

Author: I replaced it with the following sentence:
1. Download and install the latest GraalVM JDK with Native Image using [GraalVM JDK Downloader](https://github.com/graalvm/graalvm-jdk-downloader):
`bash <(curl -sL https://get.graalvm.org/jdk)`

  ```bash
  export GRAALVM_HOME=/home/${current_user}/path/to/graalvm
  export PATH=$GRAALVM_HOME/bin:$PATH
  ```
  On macOS:
  ```bash
  export GRAALVM_HOME=/Users/${current_user}/path/to/graalvm/Contents/Home
  export PATH=$GRAALVM_HOME/bin:$PATH
  ```
  On Windows:
  ```bash
  setx /M GRAALVM_HOME "C:\Progra~1\Java\<graalvm>"
  setx /M PATH "C:\Progra~1\Java\<graalvm>\bin;%PATH%"
  ```

2. Install [Native Image](https://www.graalvm.org/docs/reference-manual/native-image/#install-native-image):

  ```bash
  gu install native-image
  ```

> 3. Download or clone the repository and navigate into the `native-image-logging-examples` directory:

Reviewer 3: ```suggestion
3. Clone the repository (or download it) and change directory into the _native-image-logging-examples_ directory:
```

  ```bash
  git clone https://github.com/graalvm/graalvm-demos
  cd graalvm-demos/native-image-logging-examples
  ```

> There are two Java classes, one for the build-time logger initialization and second for runtime logger initialization. The logger will be initialized with a custom _logging.properties_ configuration file, which is placed in the same directory as _BuildTimeLoggerInit.java_ and _RuntimeLoggerInit.java_.

Reviewer 2: Update the sentence to: 

There are two Java classes: one for the build-time logger initialization and the second for runtime logger initialization.

## Build-Time Logger Initialization

In this example, the logger will be initialized at build time with a custom _logging.properties_ configuration file, placed in the same repository as _BuildTimeLoggerInit.java_.

1. Compile _BuildTimeLoggerInit.java_ using `javac`:

    ```bash
    javac BuildTimeLoggerInit.java
    ```
2. Build and run the native executable:

    ```bash
     native-image-logging-examples % native-image BuildTimeLoggerInit --initialize-at-build-time=BuildTimeLoggerInit
     ```
     ```bash
     ./buildtimeloggerinit
     ```

>      It should produce output that looks similar to:

Reviewer 2: It should produce **the** output that looks similar to:

     ```bash
     WARNING: Danger, Will Robinson! [Wed May 18 17:22:40 BST 2022]
     ```

> The _logging.properties_ file is processed at when the executable is built. `LoggerHolder.LOGGER` is initialized at build time and is available at runtime, therefore improving the startup time. Unless your application needs to process a custom _logging.properties_ configuration file at runtime, this approach is recommended.

Reviewer 2: processed **at when** the executable is built  -> processed **when** the executable is built

## Runtime Logger Initialization

The logger can also be initialized at runtime. 

1. Compile _RuntimeLoggerInit.java_ using `javac`:

     ```bash
     javac RuntimeLoggerInit.java
     ```

2. Build and run the native executable:
     ```bash
     native-image RuntimeLoggerInit -H:IncludeResources="logging.properties"
     ```
     ```bash
     ./runtimeloggerinit
     ```

    It should produce output that looks similar to:
     ```bash
     WARNING: Danger, Will Robinson! [Wed May 18 17:22:40 BST 2022]
     ```

>     In this case, the _logging.properties_ file needs to be available for runtime processing and it must be included in the executable via the `-H:IncludeResources=logging.properties` option. For more details, see [Use of Resources in a Native Executable](https://www.graalvm.org/reference-manual/native-image/dynamic-features/Resources/).

Reviewer 2: Add some conclusions at the end of the README:

Native Image supports logging using the `java.util.logging.*` API.
The logging configuration by default is based on the _logging.properties_ file found in the JDK.
